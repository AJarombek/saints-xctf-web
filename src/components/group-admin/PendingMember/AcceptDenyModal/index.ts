import AcceptDenyModal from './AcceptDenyModal';

export default AcceptDenyModal;
