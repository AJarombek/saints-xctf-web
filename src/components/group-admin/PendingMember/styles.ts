import { FontMixins } from '../../../styles/mixins';
import Colors from '../../../styles/colors';

/**
 * JSS styles for the PendingMember component.
 * @author Andrew Jarombek
 * @since 1/17/2021
 */

export default {
  pendingMember: {
    display: 'flex',
    alignItems: 'center',
    marginBottom: 10,
    borderRadius: 4,
    boxShadow: 'rgba(170, 170, 170, 0.2) 0px 1px 3px',
    backgroundColor: Colors.lightestBackground
  },
  pendingMemberActions: {
    display: 'flex',
    alignItems: 'center',
    margin: '0 10px 0 auto',

    '& > div': {
      padding: '5px 10px',
      display: 'flex'
    }
  },
  name: {
    ...FontMixins.robotoSlab(),
    fontSize: 16,
    marginLeft: 15
  },
  '@media screen and (max-width: 450px)': {
    name: {
      fontSize: 14
    },
    pendingMemberActions: {
      '& button': {
        fontSize: 12
      }
    }
  }
};
