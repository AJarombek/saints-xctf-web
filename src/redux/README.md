### Overview

Configure Redux as the state management system for the application.

### Files

| Filename                 | Description                                                                |
|--------------------------|----------------------------------------------------------------------------|
| `modules`                | Redux modules containing Reducers, Actions, and Action Creators.           |
| `appState.js`            | Sample application state maintained by Redux.                              |
| `store.js`               | Configuration and creation of the Redux store.                             |
| `types.ts`               | Typescript types used throughout the application.                          |

### Resources

1) [React/Redux Example App #1](https://github.com/erikras/react-redux-universal-hot-example/tree/master/src/redux)
2) [React/Redux Example App #2](https://github.com/goopscoop/ga-react-tutorial/tree/6-reduxActionsAndReducers/src/redux)
3) [React/Redux Example App #3](https://github.com/gothinkster/react-redux-realworld-example-app/tree/master/src)
4) [Redux Ducks Spec](https://github.com/erikras/ducks-modular-redux)
5) [React Router/Redux Setup Steps](https://github.com/supasate/connected-react-router#usage)
6) [Redux Hot Reloading](https://redux.js.org/recipes/configuring-your-store/#hot-reloading)
7) [Redux Thunk](https://github.com/reduxjs/redux-thunk)
8) [Redux/React Official Real World Example](https://github.com/reduxjs/redux/tree/master/examples/real-world)
