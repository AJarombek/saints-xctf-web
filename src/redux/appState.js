/**
 * Sample Redux State for the entire application.
 * @author Andrew Jarombek
 * @since 5/3/2020
 */

const appState = {
  auth: {
    auth: {
      isFetching: false,
      lastUpdated: 1588530309,
      signedIn: true,
      status: "SUCCESS"
    },
    registration: {
      stage: 1,
      first: "Andy",
      last: "Jarombek",
      email: "andrew@jarombek.com",
      username: undefined,
      password: undefined,
      activationCode: undefined,
      teams: {
        saintsXCTFAlumni: {
          status: "accepted",
          groups: {
            alumni: {
              status: "accepted"
            }
          }
        }
      }
    },
    user: {
      andy: {
        isFetching: false,
        didInvalidate: false,
        lastUpdated: 1588530310,
        username: "andy",
        first: "Andrew",
        last: "Jarombek",
        password: "xxxxxx",
        description: "Test Description",
        memberSince: "2016-02-23 12:00:00",
        classYear: 2017,
        location: "New York, NY",
        favoriteEvent: "Cooldown Jog",
        activationCode: "ABCD1234",
        email: "andrew@jarombek.com",
        lastSignIn: "2020-05-03 12:00:00",
        subscribed: "Y",
        logs: []
      }
    },
  },
  profile: {
    tom: {
      isFetching: false,
      didInvalidate: false,
      lastUpdated: 1588530311
    },
    joe: {
      isFetching: false,
      didInvalidate: false,
      lastUpdated: 1588530312
    },
    ben: {
      isFetching: false,
      didInvalidate: false,
      lastUpdated: 1588530313
    }
  },
  group: {
    wmenstf: {
      isFetching: false,
      didInvalidate: false,
      lastUpdated: 1588530314,
      groupName: "wmenstf",
      groupTitle: "Women's Track & Field",
      description: "",
      weekStart: "monday",
      members: [],
      logs: []
    }
  },
  logs: {
    isFetching: false,
    didInvalidate: false,
    lastUpdated: 1588530315,
    items: {
      1: {
        log_id: 1,
        username: "andy",
        first: "Andrew",
        last: "Jarombek",
        name: "Reservoir Loop",
        location: "New York, NY",
        date: "2020-05-03",
        type: "run",
        distance: 6.37,
        metric: "miles",
        miles: 6.37,
        time: "00:45:37",
        pace: "00:07:10",
        feel: 6,
        description: "",
        timeCreated: "2020-05-03 12:00:00",
        comments: {
          1: {
            commentId: 1,
            username: "andy",
            first: "Andrew",
            last: "Jarombek",
            logId: 1,
            time: "2020-05-03 12:00:00",
            content: "test"
          }
        }
      }
    }
  }
};
