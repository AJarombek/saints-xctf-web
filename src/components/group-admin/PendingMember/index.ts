import PendingMember from './PendingMember';

export default PendingMember;
