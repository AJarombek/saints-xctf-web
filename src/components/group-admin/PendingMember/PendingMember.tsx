/**
 * Component to manage the users in a group.  Admins have the ability to remove users, make users admins, and accept
 * new users.
 * @author Andrew Jarombek
 * @since 1/16/2021
 */

import React, { useState } from 'react';
import { createUseStyles } from 'react-jss';
import styles from './styles';
import { MemberDetails } from '../../../redux/types';
import { AJButton } from 'jarombek-react-components';
import { useDispatch } from 'react-redux';
import { getGroupMembers } from '../../../redux/modules/groups';
import DefaultErrorPopup from '../../shared/DefaultErrorPopup';
import AcceptDenyModal from './AcceptDenyModal';

interface Props {
  member: MemberDetails;
  groupId: number;
}

const useStyles = createUseStyles(styles);

const PendingMember: React.FunctionComponent<Props> = ({ member, groupId }) => {
  const classes = useStyles();

  const dispatch = useDispatch();

  const [showAcceptDenyModal, setShowAcceptDenyModal] = useState(false);
  const [isDenying, setIsDenying] = useState(false);
  const [isAccepting, setIsAccepting] = useState(false);
  const [inProgress, setInProgress] = useState(false);
  const [errorAcceptingMembership, setErrorAcceptingMembership] = useState(false);
  const [errorDenyingMembership, setErrorDenyingMembership] = useState(false);

  const onDeny = async (): Promise<void> => {
    setIsDenying(true);
    setInProgress(true);

    const result = dispatch(null);

    setShowAcceptDenyModal(false);
    setIsDenying(false);
    setInProgress(false);

    if (!result) {
      setErrorDenyingMembership(true);
    } else {
      dispatch(getGroupMembers(groupId));
    }
  };

  const onAccept = async (): Promise<void> => {
    setIsAccepting(true);
    setInProgress(true);

    const result = dispatch(null);

    setShowAcceptDenyModal(false);
    setIsAccepting(false);
    setInProgress(false);

    if (!result) {
      setErrorAcceptingMembership(true);
    } else {
      dispatch(getGroupMembers(groupId));
    }
  };

  return (
    <>
      <div key={member.username} className={classes.pendingMember}>
        <p className={classes.name}>
          {member.first} {member.last}
        </p>
        <div className={classes.pendingMemberActions}>
          <AJButton type="text" onClick={onDeny} disabled={false}>
            Deny
          </AJButton>
          <AJButton type="contained" onClick={onAccept} disabled={false}>
            Accept
          </AJButton>
        </div>
      </div>
      <AcceptDenyModal
        action={isAccepting ? 'accept' : isDenying ? 'deny' : null}
        onClose={(): void => setShowAcceptDenyModal(false)}
        onAccept={onAccept}
        onDeny={onDeny}
        show={showAcceptDenyModal}
        inProgress={inProgress}
        member={member}
        groupId={groupId}
      />
      {errorAcceptingMembership && (
        <DefaultErrorPopup
          message="An unexpected error occurred while accepting a users membership request.  Please try again."
          onClose={(): void => setErrorAcceptingMembership(false)}
        />
      )}
      {errorDenyingMembership && (
        <DefaultErrorPopup
          message="An unexpected error occurred while denying a users membership request.  Please try again."
          onClose={(): void => setErrorDenyingMembership(false)}
        />
      )}
    </>
  );
};

export default PendingMember;
