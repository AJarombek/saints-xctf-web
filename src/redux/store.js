/**
 * Build and configure the redux store, which holds the application state and handles
 * all state updates.
 * @author Andrew Jarombek
 * @since 5/1/2020
 */

import { createBrowserHistory } from 'history';
import { createStore, applyMiddleware } from 'redux';
import { routerMiddleware } from 'connected-react-router';
import createLogger from 'redux-logger';
import reducer from './modules/reducers';

const history = createBrowserHistory();

const configureStore = (data) => {
  const loggerMiddleware = createLogger();
  const middleware = [loggerMiddleware, routerMiddleware(history)];

  const createStoreWithMiddleware = applyMiddleware(...middleware)(createStore);
  return createStoreWithMiddleware(reducer, data);
};

export { history };
export default configureStore;
