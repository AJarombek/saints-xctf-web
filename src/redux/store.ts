/**
 * Build and configure the redux store, which holds the application state and handles
 * all state updates.
 * @author Andrew Jarombek
 * @since 5/1/2020
 */

import { createBrowserHistory } from 'history';
import { createStore, applyMiddleware, Store, Action } from 'redux';
import { routerMiddleware } from 'connected-react-router';
import { createLogger } from 'redux-logger';
import thunkMiddleware, { ThunkAction } from 'redux-thunk';
import reducer from './modules/reducers';
import { RootState } from './types';

export const history = createBrowserHistory();

export default function configureStore(): Store {
  const loggerMiddleware = createLogger();
  const middleware = [routerMiddleware(history), loggerMiddleware, thunkMiddleware];

  const createStoreWithMiddleware = applyMiddleware(...middleware)(createStore);
  const store = createStoreWithMiddleware(reducer(history));

  if (process.env.NODE_ENV !== 'production' && module.hot) {
    module.hot.accept('./modules/reducers', () => store.replaceReducer(reducer));
  }

  return store;
}

export type AppThunk<ReturnType = void, State = RootState> = ThunkAction<ReturnType, State, unknown, Action<string>>;
