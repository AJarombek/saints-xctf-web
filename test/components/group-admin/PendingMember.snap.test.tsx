/**
 * Snapshot test for the PendingMember component.
 * @author Andrew Jarombek
 * @since 3/5/2021
 */

import React from 'react';
import renderer from 'react-test-renderer';
import { Store } from 'redux';
import { emptyStore } from '../../test-utils/storeMocks';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import PendingMember from '../../../src/components/group-admin/PendingMember';

const mockStore = configureStore([thunk]);

describe('PendingMember Snapshot Tests', () => {
  let store: Store;

  beforeEach(() => {
    store = mockStore(emptyStore);
  });

  it('renders correctly', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <PendingMember
            member={{
              username: 'andy',
              first: 'Andy',
              last: 'Jarombek',
              member_since: '2016-12-23',
              user: 'admin',
              status: 'pending'
            }}
            groupId={1}
          />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
