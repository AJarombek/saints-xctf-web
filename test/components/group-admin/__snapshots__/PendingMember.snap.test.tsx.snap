// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PendingMember Snapshot Tests renders correctly 1`] = `
<div
  className="pendingMember-0-1-1"
>
  <p
    className="name-0-1-3"
  >
    Andy
     
    Jarombek
  </p>
  <div
    className="pendingMemberActions-0-1-2"
  >
    <div
      className="ajBaseButton-2-2-4 ajBaseButtonEnabled-2-2-5 aj-text-button ajTextButton-2-2-1 ajTextButtonEnabled-2-2-2"
      onClick={[Function]}
    >
      <button
        disabled={false}
        type="button"
      >
        Deny
      </button>
    </div>
    <div
      className="ajBaseButton-2-2-4 ajBaseButtonEnabled-2-2-5 aj-contained-button ajContainedButton-2-2-7 ajContainedButtonEnabled-2-2-8"
      onClick={[Function]}
    >
      <button
        disabled={false}
        type="button"
      >
        Accept
      </button>
    </div>
  </div>
</div>
`;
